### ISSUE TITLE
Implement OpenShift/Kubernetes as a backend

### ISSUE THREAD

**Issue by TomasTomecek** [user-cont/conu#155]
With current feature set, you can easily write some python code to validate your standalone containers (which run in dockerd). That's not enough. You usually run your containerized application in production orchestrated environment - Kubernetes or OpenShift. Therefore we should add support for these systems in conu.

The biggest problem with the implementation is the design itself - k8s orchestration doesn't utilize concept of containers nor images.

## Detailed plan

 * Design
   * As I wrote in the introduction, the design is the most critical part. We, as upstream maintainers, need to approve the proposed design.
   * The initial design is expected to be simple: conu needs to be able to run a containerized application inside k8s. For that, we need only a few k8s primitives to be implemented:
     * pod, service, deployment
   * In case of OpenShift, the app needs to be runnable via `oc new-app`.
   * We also require convenient user experience. Ideally users would provide well-known configuration object without the need of learning some complex python API, an example:
     ```python
     from conu import Pod
     pod_spec = """
         apiVersion: v1
         kind: Pod
         metadata:
           name: my-pod
         spec:
           containers:
           - env:
             - name: SOME_ENV_VAR
               value: "I am a value"
             image: library/fedora:27
             imagePullPolicy: IfNotPresent
             name: my-container
           restartPolicy: Never
     """
     pod = Pod(from_spec=pod_spec)
     # run the pod now
     ```
 * Implementation
   * Once the design is approved, you can start working on the implementation.
 * It is expected to iterate on the design and implementation phases multiple times so we can achieve even the more complex scenarios.
 * Configuration
   * Orchestration systems are very complex. When interacting with them, users usually require a flexible workflow which can be customized.
   * Users should be able to specify: cluster, API token, the way to connect, namespace, user, clean after testing? And even more!
 * Template support
   * Once we have the basics, it's time to support templates -- this is the usual way people deploy to k8s/OpenShift.
 * Parametrizing templates
    * In order to be flexible and allow parametrizing test cases, we may need to be able to parametrize templates -- users can easily change parts of a template via some well-known templating system (such as jinja).
 * Finish line
    * To prove that all the new code works, use it to write at least 2 simple test suites for container images delivered in Fedora and validate they work well in k8s/openshift.

This issue [is proposed](https://docs.fedoraproject.org/mentored-projects/gsoc/2018/ideas.html#complete-support-for-openshift-and-kubernetes-as-a-backend-in-conu) as a GSoC project as part of Fedora.

**Comment by mateimicu** [comment]
Hi, 
I'm interested in this idea for GSoC but I want to understand more.

From my understanding, this should implement the k8s interface as a backend so you can plug the configurations for an existing k8s deployment and then run the Pod from the yaml definition. 

My question is if we should also write wrappers/utils for different functions/actions you can do on pods(for example)? I'm interested in how much should the abstraction provide?
With the new concepts (pods, deployments ... etc ) you have new actions that you can perform, will the `conu` project support this actions or just a subset and allow the user to use for example the k8s python client for something else more specific? 

Another question is about the yaml definition: 
We can integrate jinja easily in the template but do we need to process it more? I would assume is enough to render the template and then pass it to the k8s cluster to create the resource.  Maybe some type checking and linting to see if it's a valid yaml but letting the k8s parser to see if the string is valid.

It looks to the that the design will make/break this type of backend, should the application contain the "final" design or the actual design can be iterated as part of the GSoC program? I'm asking this because coming up with a viable design (even in draft form) might take some time and I don't know if I can get one done in time for the application process.

Thanks in advance! :D

**Comment by TomasTomecek** [comment]
Hi Matei, thanks for your interest!

The overall comment I have for you is that the design should be solved as part of the GSoC project.

> My question is if we should also write wrappers/utils for different functions/actions you can do on pods(for > example)?

I would say that yes. The main intent here is to provide this interface to people so they can use it to validate their containerized applications running in kube. I think that implementation for pods will have a smaller prio than implementation for Services and Deployments.

> With the new concepts (pods, deployments ... etc ) you have new actions that you can perform, will the
> conu project support this actions or just a subset and allow the user to use for example the k8s python
> client for something else more specific?

Ideally the implementation would use that API client and it would be fully available to users.

> We can integrate jinja easily in the template but do we need to process it more? I would assume is enough 
> to render the template and then pass it to the k8s cluster to create the resource.

Yup.

> Maybe some type checking and linting to see if it's a valid yaml but letting the k8s parser to see if the 
> string is valid.

If users want to shoot themselves to foot, they are free to do so. Type checking and linting is definitely a nice to have feature.

> It looks to the that the design will make/break this type of backend, should the application contain the 
> "final" design or the actual design can be iterated as part of the GSoC program? I'm asking this because
> coming up with a viable design (even in draft form) might take some time and I don't know if I can get one
> done in time for the application process.

I completely share your concern. The key here is not to overthink the design and instead start small, iterate and converge to a shared goal -- see last point in the list.

**Comment by dhodovsk** [comment]
GSoC finished succesfully and the work has been done by @rpitonak
